src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,284 @@
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public boolean remove(T object) {
+        if (head == null || isEmpty()) {
+            return false;
+        }
+        int index = indexOf(object);
+        if (index != -1) {
+            remove(index);
+            return true;
+        } else {
+            return false;
+        }
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public T remove(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(INDEX_OF_BOUND_EXCEPTION_MESSAGE);
+        }
+        T deletedValue;
+        if (size == 1) {
+            deletedValue = head.value;
+            removeLastElement();
+        } else if (index == 0) {
+            deletedValue = head.value;
+            removeElementFromBeginning();
+        } else if (index == size - 1) {
+            deletedValue = tail.value;
+            removeElementFromEnd();
+        } else {
+            deletedValue = removeFromMiddle(index);
+        }
+        size--;
+        return deletedValue;
+    }
+
+    private void removeElementFromEnd() {
+        tail = tail.prev;
+        tail.next = null;
+    }
+
+    private void removeElementFromBeginning() {
+        head = head.next;
+        head.prev = null;
+    }
+
+    private void removeLastElement() {
+        head = null;
+        tail = null;
+    }
+
+    private T removeFromMiddle(int index) {
+        Node<T> current = head;
+        int counter = 0;
+        int step = 1;
+        String iterationVariant = FROM_START_TYPE_OF_ITERATION;
+        if (index > size / 2) {
+            current = tail;
+            counter = size - 1;
+            step = -1;
+            iterationVariant = FROM_END_TYPE_OF_ITERATION;
+        }
+        while (current != null) {
+            if (counter == index) {
+                break;
+            }
+            current = iterationVariant.equals(FROM_START_TYPE_OF_ITERATION)
+                    ? current.next : current.prev;
+            counter += step;
+        }
+        unlink(current);
+        return current.value;
+    }
+
+    private void unlink(Node<T> current) {
+        if (current != null) {
+            Node<T> newNext = current.next;
+            Node<T> newPrev = current.prev;
+            newNext.prev = newPrev;
+            newPrev.next = newNext;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public int indexOf(T o) {
+        int counter = 0;
+        Node<T> current = head;
+        while (current != null) {
+            if (equals(current.value, o)) {
+                return counter;
+            }
+            counter++;
+            current = current.next;
+        }
+        return -1;
+    }
+
+    public boolean equals(T first, T second) {
```

Comment:
what is this equals? what are you comparing? why is it public?

Reply:
make equals private. This quals check if two objects are equals even if them are null. The same thing did Objects,equals(a,b)